File: app/src/main/java/ru/art2000/calculator/view/currency/CurrenciesEditFragment.java
package ru.art2000.calculator.view.currency;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import kotlin.Unit;
import ru.art2000.calculator.R;
import ru.art2000.calculator.databinding.ModifyCurrenciesLayoutBinding;
import ru.art2000.calculator.model.currency.CurrencyItem;
import ru.art2000.calculator.view_model.currency.CurrenciesEditModel;
import ru.art2000.calculator.view_model.currency.CurrenciesSettingsModel;
import ru.art2000.calculator.view_model.currency.CurrencyDependencies;
import ru.art2000.extensions.collections.CollectionsKt;
import ru.art2000.extensions.collections.LiveList;
import ru.art2000.extensions.fragments.UniqueReplaceableFragment;
import ru.art2000.extensions.views.ViewsKt;
import ru.art2000.helpers.AndroidHelper;

public class CurrenciesEditFragment extends UniqueReplaceableFragment {

    private ItemTouchHelper itemTouchHelper;

    private ModifyCurrenciesLayoutBinding viewBinding;
    private CurrenciesEditModel model;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        if (viewBinding == null) {

            model = new ViewModelProvider(requireActivity()).get(CurrenciesSettingsModel.class);

            viewBinding = ModifyCurrenciesLayoutBinding.inflate(inflater);

            viewBinding.modifyCurrenciesList.setPadding(0, 0, 0, AndroidHelper.dip2px(requireContext(), 20));

            viewBinding.modifyCurrenciesList.setEmptyViewGenerator((context, viewGroup, integer) ->
                    ViewsKt.createTextEmptyView(context, getEmptyTextRes()));

            EditCurrenciesAdapter adapter = new EditCurrenciesAdapter();
            viewBinding.modifyCurrenciesList.setAdapter(adapter);

            LinearLayoutManager llm = new LinearLayoutManager(requireContext());
            llm.setOrientation(RecyclerView.VERTICAL);
            viewBinding.modifyCurrenciesList.setLayoutManager(llm);

            itemTouchHelper = new ItemTouchHelper(new CurrenciesEditRecyclerTouchCallback(
                    requireContext(),
                    position -> {

                        CurrencyItem removedItem = model.getDisplayedVisibleItems().get(position);
                        model.databaseMarkHidden(removedItem);

                        return Unit.INSTANCE;
                    },
                    (firstPosition, secondPosition) -> {
                        adapter.swap(firstPosition, secondPosition);
                        return Unit.INSTANCE;
                    }
            ));
            itemTouchHelper.attachToRecyclerView(viewBinding.modifyCurrenciesList);
        }

        return viewBinding.getRoot();
    }

    @Override
    public void onReselected() {
        viewBinding.modifyCurrenciesList.smoothScrollToPosition(0);
    }

    @StringRes
    private int getEmptyTextRes() {
        return R.string.empty_text_no_currencies_added;
    }

    @Override
    public int getOrder() {
        return 1;
    }

    @Override
    public int getTitle() {
        return R.string.currencies_edit;
    }

    private class EditCurrenciesAdapter extends RecyclerView.Adapter<EditCurrenciesAdapter.Holder> {

        private final int REORDER_MODE = 0;
        private final int SELECTION_MODE = 1;

        @LayoutRes
        private final int selectionItem = R.layout.item_add_currencies_list;
        @LayoutRes
        private final int reorderItem = R.layout.item_edit_currencies_list;
        private int curMode = REORDER_MODE;

        EditCurrenciesAdapter() {
            model.getSelectedVisibleItems().observe(getViewLifecycleOwner(), new LiveList.LiveListObserver<CurrencyItem>() {

                @Override
                public void onItemsInserted(@NotNull List<? extends CurrencyItem> previousList,
                                            @NotNull List<? extends CurrencyItem> insertedItems,
                                            int position) {
                    super.onItemsInserted(previousList, insertedItems, position);

                    for (CurrencyItem item : insertedItems) {
                        markItem(item, true);
                    }
                }

                @Override
                public void onItemsRemoved(@NotNull List<? extends CurrencyItem> previousList,
                                           @NotNull List<Integer> removedItems) {
                    super.onItemsRemoved(previousList, removedItems);
                    if (model.getSelectedVisibleItems().isEmpty()) {
                        setReorderMode();
                    }

                    for (int i : removedItems) {
                        markItem(previousList.get(i), false);
                    }
                }
            });

            model.getDisplayedVisibleItems().observe(getViewLifecycleOwner(), new LiveList.LiveListObserver<CurrencyItem>() {
                @Override
                public void onAnyChanged(@NotNull List<? extends CurrencyItem> previousList) {
                    dispatchListUpdate(previousList, model.getDisplayedVisibleItems());
                }
            });
        }

        @Override
        public int getItemViewType(int position) {
            return curMode;
        }

        @NonNull
        @Override
        public EditCurrenciesAdapter.Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View item;
            if (viewType == SELECTION_MODE)
                item = LayoutInflater.from(requireContext()).inflate(selectionItem, null);
            else
                item = LayoutInflater.from(requireContext()).inflate(reorderItem, null);
            return new Holder(item);
        }

        @Override
        public void onBindViewHolder(@NonNull EditCurrenciesAdapter.Holder holder, int position) {
            CurrencyItem currencyItem = model.getDisplayedVisibleItems().get(position);
            holder.bind(currencyItem);
        }

        @Override
        public int getItemCount() {
            return model.getDisplayedVisibleItems().size();
        }

        void swap(int position, int anotherPosition) {
            if (position < 0
                    || anotherPosition < 0
                    || position >= getItemCount()
                    || anotherPosition >= getItemCount()) {

                return;
            }

            CurrencyItem item = model.getDisplayedVisibleItems().get(position);
            CurrencyItem anotherItem = model.getDisplayedVisibleItems().get(anotherPosition);

            item.position = anotherPosition;
            anotherItem.position = position;

            Map<CurrencyItem, CurrencyItem> map = new HashMap<>();
            map.put(item, anotherItem);
            map.put(anotherItem, item);

            model.getDisplayedVisibleItems().replaceAll(map);
        }

        void dispatchListUpdate(List<? extends CurrencyItem> oldData, List<? extends CurrencyItem> newData) {
            DiffUtil.DiffResult result = CollectionsKt.calculateDiff(oldData, newData);
            result.dispatchUpdatesTo(this);
        }

        private void setReorderMode() {
            if (curMode == REORDER_MODE)
                return;

            curMode = REORDER_MODE;
            model.setEditSelectionMode(false);

            itemTouchHelper.attachToRecyclerView(viewBinding.modifyCurrenciesList);
            model.getSelectedVisibleItems().clear();
        }

        private void setSelectionMode(RecyclerView.ViewHolder holder) {
            curMode = SELECTION_MODE;
            model.setEditSelectionMode(true);

            itemTouchHelper.attachToRecyclerView(null);
            model.dismissFirstTimeTooltip();

            int p = holder.getBindingAdapterPosition();
            CurrencyItem item = model.getDisplayedVisibleItems().get(p);
            model.setVisibleItemSelected(item, true);
            notifyDataSetChanged();
        }

        private void markItem(CurrencyItem currencyItem, boolean selected) {
            Holder holder = (Holder)
                    viewBinding.modifyCurrenciesList.findViewHolderForAdapterPosition(
                            model.getDisplayedVisibleItems().indexOf(currencyItem));

            if (holder == null || holder.checkBox == null)
                return;

            holder.checkBox.setChecked(selected);
        }

        class Holder extends RecyclerView.ViewHolder {

            TextView code;
            ImageView handle;
            CheckBox checkBox;
            TextView name;

            Holder(final View itemView) {
                super(itemView);
                code = itemView.findViewById(R.id.currency_code);
                name = itemView.findViewById(R.id.currency_name);
                handle = itemView.findViewById(R.id.handle);
                checkBox = itemView.findViewById(R.id.checkbox_add);
            }

            @SuppressLint("ClickableViewAccessibility")
            void bind(CurrencyItem currencyItem) {
                code.setText(currencyItem.code);

                name.setText(CurrencyDependencies.getNameIdentifierForCode(requireContext(), currencyItem.code));

                if (handle != null) {
                    handle.setOnTouchListener((v, event) -> {
                        if (event.getActionMasked() == MotionEvent.ACTION_DOWN) {
                            itemTouchHelper.startDrag(this);
                        }
                        return false;
                    });
                    itemView.setOnLongClickListener(v -> {
                        setSelectionMode(this);
                        return false;
                    });
                } else {
                    checkBox.setOnCheckedChangeListener(null);
                    checkBox.setChecked(model.isVisibleItemSelected(currencyItem));
                    checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
                        CurrencyItem item = model.getDisplayedVisibleItems().get(getBindingAdapterPosition());
                        if (isChecked != model.isVisibleItemSelected(item)) {
                            model.setVisibleItemSelected(item, isChecked);
                        }
                    });
                    itemView.setOnClickListener(v -> checkBox.performClick());
                }
            }
        }
    }
}

File: app/src/main/java/ru/art2000/calculator/view/currency/CurrenciesEditFragment.kt
package ru.art2000.calculator.view.currency

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import by.kirich1409.viewbindingdelegate.CreateMethod
import by.kirich1409.viewbindingdelegate.viewBinding
import ru.art2000.calculator.R
import ru.art2000.calculator.databinding.ItemAddCurrenciesListBinding
import ru.art2000.calculator.databinding.ItemEditCurrenciesListBinding
import ru.art2000.calculator.databinding.ModifyCurrenciesLayoutBinding
import ru.art2000.calculator.model.currency.CurrencyItem
import ru.art2000.calculator.model.currency.getNameIdentifier
import ru.art2000.calculator.view_model.currency.CurrenciesEditModel
import ru.art2000.calculator.view_model.currency.CurrenciesSettingsModel
import ru.art2000.extensions.collections.LiveList
import ru.art2000.extensions.collections.LiveList.LiveListObserver
import ru.art2000.extensions.collections.calculateDiff
import ru.art2000.extensions.fragments.UniqueReplaceableFragment
import ru.art2000.extensions.views.createTextEmptyView
import ru.art2000.helpers.dip2px

class CurrenciesEditFragment : UniqueReplaceableFragment() {

    private var itemTouchHelper: ItemTouchHelper? = null
    private val binding by viewBinding<ModifyCurrenciesLayoutBinding>(CreateMethod.INFLATE)
    private val model: CurrenciesEditModel by activityViewModels<CurrenciesSettingsModel>()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {

        binding.modifyCurrenciesList.setPadding(0, 0, 0, requireContext().dip2px(20f))
        binding.modifyCurrenciesList.emptyViewGenerator = { ctx, _, _ ->
            createTextEmptyView(ctx, emptyTextRes)
        }
        val adapter = EditCurrenciesAdapter()
        binding.modifyCurrenciesList.adapter = adapter
        val llm = LinearLayoutManager(requireContext())
        llm.orientation = RecyclerView.VERTICAL
        binding.modifyCurrenciesList.layoutManager = llm
        itemTouchHelper = ItemTouchHelper(CurrenciesEditRecyclerTouchCallback(
            requireContext(),
            { position ->
                val removedItem = model.displayedVisibleItems[position]
                model.databaseMarkHidden(removedItem)
            }
        ) { firstPosition, secondPosition ->
            adapter.swap(firstPosition, secondPosition)
        }).apply {
            attachToRecyclerView(binding.modifyCurrenciesList)
        }

        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        itemTouchHelper = null
    }

    override fun onReselected() {
        binding.modifyCurrenciesList.smoothScrollToPosition(0)
    }

    @get:StringRes
    private val emptyTextRes: Int
        get() = R.string.empty_text_no_currencies_added

    override fun getTitle(): Int {
        return R.string.currencies_edit
    }

    companion object {
        private const val REORDER_MODE = 0
        private const val SELECTION_MODE = 1
    }

    private inner class EditCurrenciesAdapter :
        RecyclerView.Adapter<EditCurrenciesAdapter.Holder>() {

        private var curMode = REORDER_MODE

        init {
            model.selectedVisibleItems.observe(
                viewLifecycleOwner,
                object : LiveListObserver<CurrencyItem>() {
                    override fun onItemsInserted(
                        previousList: List<CurrencyItem>,
                        liveList: LiveList<CurrencyItem>,
                        insertedIndices: List<Int>
                    ) {
                        super.onItemsInserted(previousList, liveList, insertedIndices)
                        for (i in insertedIndices) {
                            markItem(liveList[i], true)
                        }
                    }

                    override fun onItemsRemoved(
                        previousList: List<CurrencyItem>,
                        removedIndices: List<Int>
                    ) {
                        super.onItemsRemoved(previousList, removedIndices)
                        if (model.selectedVisibleItems.isEmpty()) {
                            setReorderMode()
                        }
                        for (i in removedIndices) {
                            markItem(previousList[i], false)
                        }
                    }
                })
            model.displayedVisibleItems.observe(
                viewLifecycleOwner,
                object : LiveListObserver<CurrencyItem>() {
                    override fun onAnyChanged(previousList: List<CurrencyItem>) {
                        dispatchListUpdate(previousList, model.displayedVisibleItems)
                    }
                })
        }

        override fun getItemViewType(position: Int) = curMode

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
            val inflater = LayoutInflater.from(requireContext())
            return when (viewType) {
                SELECTION_MODE -> Holder(
                    ItemAddCurrenciesListBinding.inflate(
                        inflater,
                        parent,
                        false
                    )
                )
                REORDER_MODE -> Holder(
                    ItemEditCurrenciesListBinding.inflate(
                        inflater,
                        parent,
                        false
                    )
                )
                else -> throw IllegalArgumentException("Unknown viewType '$viewType'")
            }
        }

        override fun onBindViewHolder(holder: Holder, position: Int) {
            val currencyItem = model.displayedVisibleItems[position]
            holder.bind(currencyItem)
        }

        override fun getItemCount() = model.displayedVisibleItems.size

        fun swap(position: Int, anotherPosition: Int) {
            if (position < 0 || anotherPosition < 0 || position >= itemCount || anotherPosition >= itemCount) {
                return
            }
            val item = model.displayedVisibleItems[position]
            val anotherItem = model.displayedVisibleItems[anotherPosition]
            item.position = anotherPosition
            anotherItem.position = position
            val map = buildMap {
                this[item] = anotherItem
                this[anotherItem] = item
            }
            model.displayedVisibleItems.replaceAll(map)
        }

        fun dispatchListUpdate(oldData: List<CurrencyItem>, newData: List<CurrencyItem>) {
            val result = calculateDiff(oldData, newData)
            result.dispatchUpdatesTo(this)
        }

        private fun setReorderMode() {
            if (curMode == REORDER_MODE) return
            curMode = REORDER_MODE
            model.isEditSelectionMode = false
            itemTouchHelper!!.attachToRecyclerView(binding.modifyCurrenciesList)
            model.selectedVisibleItems.clear()
        }

        @SuppressLint("NotifyDataSetChanged")
        private fun setSelectionMode(holder: RecyclerView.ViewHolder) {
            curMode = SELECTION_MODE
            model.isEditSelectionMode = true
            itemTouchHelper!!.attachToRecyclerView(null)
            model.dismissFirstTimeTooltip()
            val p = holder.bindingAdapterPosition
            val item = model.displayedVisibleItems[p]
            model.setVisibleItemSelected(item, true)
            notifyDataSetChanged()
        }

        private fun markItem(currencyItem: CurrencyItem, selected: Boolean) {
            val holder = binding.modifyCurrenciesList.findViewHolderForAdapterPosition(
                model.displayedVisibleItems.indexOf(currencyItem)
            ) as Holder? ?: return
            holder.checkBox?.isChecked = selected
        }

        inner class Holder private constructor(
            root: View,
            val code: TextView,
            val name: TextView,
            val checkBox: CheckBox?,
        ) : RecyclerView.ViewHolder(root) {

            constructor(binding: ItemAddCurrenciesListBinding) : this(
                binding.root,
                binding.currencyCode, binding.currencyName,
                binding.checkboxAdd,
            ) {
                binding.checkboxAdd.apply {
                    setOnClickListener {
                        val item = model.displayedVisibleItems[bindingAdapterPosition]
                        if (isChecked != model.isVisibleItemSelected(item)) {
                            model.setVisibleItemSelected(item, isChecked)
                        }
                    }
                }

                itemView.setOnClickListener { binding.checkboxAdd.performClick() }
            }

            @SuppressLint("ClickableViewAccessibility")
            constructor(binding: ItemEditCurrenciesListBinding) : this(
                binding.root,
                binding.currencyCode, binding.currencyName,
                null,
            ) {
                binding.handle.setOnTouchListener { _, event ->
                    if (event.actionMasked == MotionEvent.ACTION_DOWN) {
                        itemTouchHelper!!.startDrag(this)
                    }
                    false
                }
                binding.root.setOnLongClickListener {
                    setSelectionMode(this)
                    false
                }
            }

            fun bind(currencyItem: CurrencyItem) {
                code.text = currencyItem.code
                name.setText(currencyItem.getNameIdentifier(requireContext()))
                checkBox?.isChecked = model.isVisibleItemSelected(currencyItem)
            }
        }
    }
}